R/coverage-matrix.R

```diff
@@ -0,0 +1,156 @@
+#' Scrape "coverage" tag in test code to generate mapping
+#' @param references dynamic reference holder if it already exists
+#' @param src,ref passed to \code{\link{vt_scrape_tags_from}}
+#' @param type one of "long" or "wide" which determines shape of output table
+#' @return a data.frame mapping requirement ids to test case ids.
+#' @importFrom rlang list2 := !!
+#' @export
+vt_scrape_coverage_matrix <- function(type = c("long", "wide"), references = NULL, src = ".", ref = vt_path()){
+
+  ## instantiate dynamic referencing
+  ## if no dynamic reference content, this won't get populated
+  if(is.null(references)){
+    references <- vt_dynamic_referencer$new()
+  }
+
+  ## helper functions
+
+  split_vals <- function(x){
+    this_row <- strsplit(x["coverage"], split = ":")[[1]]
+    names(this_row) <- c("tc_id", "req_id")
+    this_row["tc_id"] <- trimws(this_row["tc_id"])
+
+    this_row["tc_title"] <- x["tc_title"]
+    as.data.frame(t(this_row))
+  }
+
+  split_req <- function(vals){
+    do.call("rbind", apply(vals, 1, FUN = function(x){
+      req_one_row <- data.frame(tc_title = x[["tc_title"]],
+                                tc_id = x[["tc_id"]],
+                                req_id = strsplit(trimws(x[["req_id"]]), split = ", ")[[1]])
+      req_one_row$req_title <- paste0("Requirement ", gsub(req_one_row$req_id,
+                                                           pattern = "^(\\d+)\\.*.*",
+                                                           replacement = "\\1"))
+      req_one_row
+    }))
+  }
+
+
+
+  # avoids dependency on tidyr::pivot_wider
+  make_wider <- function(long_vals){
+    list_x <- apply(long_vals, 1, FUN = function(x){
+      as.data.frame(list2(req_id = x[["req_id"]],
+                          req_title = x[["req_title"]],
+                          !!x[["tc_id"]] := "x"),
+                    check.names = FALSE)
+    })
+
+    all_names <- unique(unlist(lapply(list_x, names)))
+    list_all_x <- lapply(list_x, function(x){
+      x[setdiff(all_names, names(x))] <- ""
+      x
+    })
+    out <- do.call("rbind", list_all_x)
+    row.names(out) <- 1:nrow(out)
+    out[, c("req_title", "req_id", sort(names(out)[-1:-2]))]
+  }
+
+
+  ## end helper functions
+
+  cov_raw_values <- sapply(vt_scrape_tags_from(type = "test_cases", tags =  c("title", "coverage"),
+                                               src = src, ref = vt_path()),
+                           FUN = function(x){data.frame(coverage = x$coverage,
+                                                        tc_title = x$title)})
```

Comment:
does this need to be a sapply? you could do.call('rbind' and then change the column name once...

Reply:
this will get updated as part of using `dynamic_reference_rendering` update